src/time/time.cr

```diff
@@ -567,42 +585,55 @@ struct Time
   # Time.local_offset_in_minutes # => -180
   # ```
   def self.local_offset_in_minutes
-    ifdef linux
-      -LibC.timezone.to_i32 / 60
-    else
-      if LibC.gettimeofday(nil, out tzp) != 0
-        raise Errno.new("gettimeofday")
-      end
-      -tzp.tz_minuteswest.to_i32
+    compute_ticks do |t, offset|
+      offset / Span::TicksPerMinute
     end
   end
 
   protected def self.compute_utc_ticks(ticks)
-    compute_ticks do |t, tp, tz|
-      ticks + tz
+    compute_ticks do |t, offset|
+      ticks - offset
     end
   end
 
   protected def self.compute_local_ticks(ticks)
-    compute_ticks do |t, tp, tz|
-      ticks - tz
+    compute_ticks do |t, offset|
+      ticks + offset
     end
   end
 
+  # Yields `ticks, offset`, where `ticks` is the number of ticks
+  # for "now", and `offset` is the number of ticks for now's timezone offset.
   private def self.compute_ticks
-    if LibC.gettimeofday(out tp, out tzp) != 0
-      raise Errno.new("gettimeofday")
-    end
-    ticks = tp.tv_sec.to_i64 * Span::TicksPerSecond + tp.tv_usec.to_i64 * 10_i64
-    ticks += UnixEpoch
+    LibC.tzset
 
     ifdef linux
-      tz = LibC.timezone.to_i64 / 60
+      ret = LibC.clock_gettime(LibC::CLOCK_REALTIME, out timespec)
+      raise Errno.new("clock_gettime") unless ret == 0
+      tv_sec, tv_usec = timespec.tv_sec, timespec.tv_nsec / 1_000
```

Comment:
Maybe out of scope, but how about instead of reducing precision here, use a finer precision down below (maybe even increase the precision of `ticks`?) and doing `tv_sec, tv_nsec = timeval.tv_sec, timeval.tv_usec * 1_000` on other platforms?